source/common/sagemath/library.sage

```diff
@@ -30,3 +34,332 @@ class TBIL:
         p.ymin(-1)
         p.ymax(1)
         return p
+
+    @staticmethod
+    def numberline_plot(center=0, radius=10):
+        P = arrow((center-radius,0),(center+radius,0),color="black", width=1, arrowsize=1, aspect_ratio=1,head=2)
+        
+        #Print approximately every (radius/10)-th label for longer lines
+        label_skip=1
+        if radius>10:
+            label_skip=round(radius/10)
+
+        for i in range(center-radius+1,center+radius):
+            if i % label_skip == 0:
+                P += line([(i,-0.2),(i,0.2)],color="black")
+                P += text(f"${i}$", (i,-0.6),color="black")
+        return P
+    
+    @staticmethod
+    def inequality_plot(
+            start=None, 
+            strict_start=True, 
+            end=None, 
+            strict_end=True,
+            label_endpoints=True,
+            scale=10):
+        P = Graphics()
+        if start is None:
+            P += arrow((end,0),(-1*scale,0),color="#0088ff", width=3, arrowsize=3, aspect_ratio=1)
+        if end is None:
+            P += arrow((start,0),(1*scale,0),color="#0088ff", width=3, arrowsize=3, aspect_ratio=1)
+        if start is not None and end is not None:
+            P += line([(start,0),(end,0)],color="#0088ff", thickness=3, aspect_ratio=1)
+
+        if start is not None:
+            if label_endpoints:
+                P += text(f"${round(start,ndigits=2)}$", (start,0.6), color="black")
+            if strict_start:
+                P += text("(", (start,0), color="#0088ff", fontsize=18)
+            else:
+                P += text("[", (start,0), color="#0088ff", fontsize=18)
+
+        if end is not None:
+            if label_endpoints:
+                P += text(f"${round(end,ndigits=2)}$", (end,0.6), color="black")
+            if strict_end:
+                P += text(")", (end,0), color="#0088ff", fontsize=18)
+            else:
+                P += text("]", (end,0), color="#0088ff", fontsize=18)
+        return P
+
+    @staticmethod
+    def intervals_from_inequality(inequality, partition_points,undefined_points=[],checkpoints=None):
```

Comment:
```suggestion
    def intervals_from_inequality(inequality, partition_points,undefined_points=None,checkpoints=None):
        if undefined_points is None: undefined_points = []
```

Reply:
How is this different?

Reply:
https://www.pullrequest.com/blog/python-pitfalls-the-perils-of-using-lists-and-dicts-as-default-arguments/

Reply:
Tell me again about how Python is great because the code is so readable? 🫠

(Thanks for the explainer.)

Reply:
Yeah, I'm not sure what advantage there is to assign default arguments at function instantiation rather than function call, but there it is.

Reply:
I'm glad you noticed this, because I feel like we have chased a bug like this around before and it took a while to find.